Pandas updates - sort and modify

ajkerrigan opened the pull request:
I hit an issue while trying to sort a pandas sheet (#495) and was working on it locally. While trying to avoid stomping on other related issues or previous fixes, I came across the modification issue in #360. This PR attempts to resolve both issues, though the changes are in separate commits so we can tweak or divide the PR as needed.

The modification issue appeared to stem from the fact that the pandas `read_stata` function returns a DataFrame with an Int64Index instead of the default RangeIndex. The warning during edits doesn't seem to show up when using the pandas loader with CSV/JSON/Parquet files. I dug into the difference between Int64Index and RangeIndex a bit ([reference](https://pandas.pydata.org/pandas-docs/stable/user_guide/advanced.html#int64index-and-rangeindex)), and it seems like converting Int64Index to RangeIndex could cover what `reload()` is already trying to do without forcing a `reset_index()`. There is probably a better way to handle it, but this seems like forward progress at least :)

ajkerrigan committed:
Fix errors during pandas sort

* Explicitly sidestep `DataFrameAdapter`'s `__getattr__()` override
when accessing `self.df`. This avoids `RecursionError` during the
sheet copy phase of a sort.

* Update `sort()` to handle column/reverse pairs in `self._ordering`

ajkerrigan committed:
Convert pandas Int64Index to RangeIndex

Convert a DataFrame's index from Int64Index to RangeIndex during a
reload. Since RangeIndex is a special case subset of Int64Index, this
guarantees a RangeIndex without requiring a full reset_index().

anjakefala commented on visidata/loaders/_pandas.py:
```diff
@@ -17,6 +17,9 @@ def __getitem__(self, k):
     def __getattr__(self, k):
+        # This check helps avoid infinite recursion trouble
+        if k == 'df':
+            raise AttributeError
```
```
if hasattr(y, '__setstate__'):
```

I bumped into, too!

This occurs because we do not handle missing attributes. I think I prefer we checking if self has the attribute `df`, since that captures the root of the problem, and raise AttributeError otherwise.

Like here:
https://github.com/saulpw/visidata/pull/491/files

anjakefala replied:
@saulpw, we should audit the code for all implementations of `__getattr__`, and ensure that each handles what happens when attributes are missing.

ajkerrigan replied:
True @anjakefala - calling out `df` explicitly looks hacky and I'm sure we can make it cleaner. It would probably make sense to have `DataFrameAdapter` be a `pd.DataFrame` subclass if not for the initial loading time. The way you linked could work and be more flexible (in case we need to add other local attributes someday).

anjakefala commented on visidata/loaders/_pandas.py:
```diff
@@ -67,7 +70,9 @@ def reload(self):
 
         # reset the index here
-        if type(self.df.index) is not pd.RangeIndex:
+        if type(self.df.index) is pd.Int64Index:
```
Really nice find!

Should we keep an eye out for other Indexing types?

ajkerrigan replied:
Not really sure to be honest. Converting Int64-->Range here seems clear enough here (assuming we want/need a RangeIndex), but I don't know enough to extrapolate beyond that.

anjakefala replied:
Okay, perhaps this is the best we can do for now, and we can mature it from there. I did test a few formats, and things seem to be fine. =) Perhaps folks who care about other formats can chip in if they notice something. Thanks!

anjakefala commented on visidata/loaders/_pandas.py:
```diff
@@ -90,9 +95,9 @@ def sort(self):
-        for cols, reverse in self._ordering[::-1]:
-            by_cols += [col.name for col in cols]
-            ascending += [not reverse] * len(cols)
+        for col, reverse in self._ordering[::-1]:
```
@saulpw, can `self._ordering[::]` contain multiple columns? Probably when you have multiple key columns, and do `g[`?

saulpw replied:
Yes; each item in `_ordering` should be a tuple of `(list(columns), reverse_bool)`.

anjakefala replied:
Judging from @ajkerrigan's linked issue report: https://github.com/saulpw/visidata/issues/495

It has not been a `list(columns)`, so we should look into that!

ajkerrigan replied:
I did do some tests locally with `z[`/`z]` to make sure this behaved when sorting by multiple columns. Because in those cases, `self._ordering` looks like:

```[(<Column 1>, True), (<Column 2>, False)]```

Still definitely possible that I missed a piece of behavior, but as far as I can tell sorting works with single or multiple columns.

saulpw replied:
My apologies, I changed it a few months ago in 9da019ca4, to be `(column, reverse_bool)`.  So that is the issue, that the "interface" for `_ordering` changed and the pandas loader did not change with it.  I wonder if `_ordering` should lose its private `_` and become a documented interface?  Other loaders/plugins that override provide custom sort functionality would need access to it as well, so we should probably formalize it.

anjakefala replied:
I can also confirm that it does work!

anjakefala replied:
Agreed about formalizing `_ordering`, and properly documenting it..

anjakefala commented:
This is AMAZING and WELCOME. It makes me really, really happy! :purple_heart:

saulpw reviewed (changes requested):
Thanks for figuring this out, @ajkerrigan!  Let's find the right pattern for the AttributeError raise (since we need it in a couple places like this), and apply it here.  Otherwise looks great, I think you saved us hours of debugging pandas indexes!

ajkerrigan committed:
Handle DataFrameAdapter's state more gracefully

Use __getstate__/__setstate__ to delegate a DataFrameAdapter's state to
the underlying DataFrame.

Since DataFrameAdapter is a proxy/wrapper class around a DataFrame, we
can keep a more common __getattr__/__setattr__ combination to delegate
public attribute access to the underlying DataFrame. Overriding __getstate__
and __setstate__ avoids confusion and recursion errors during pickling or
copy operations.

ajkerrigan commented:
> Thanks for figuring this out, @ajkerrigan! Let's find the right pattern for the AttributeError raise (since we need it in a couple places like this), and apply it here. Otherwise looks great, I think you saved us hours of debugging pandas indexes!

Thanks for the review! I'm with you, sorting out the attribute behavior in a repeatable way seems handy. During my local edits I experimented both with tweaks to `__getattr__` and overriding `__copy__`.

After a bit more reading, I think we can address the recursion issue in a more general way by overriding `__getstate__` and `__setstate__`. That last commit is a working sample, but it's probably worth talking through a bit. We'll get there! :)

anjakefala commented:
@ajkerrigan, I am fine with implementing `__getstate__` and `__setstate__`! I am a bit confused why that is more general than handling the potential of a missing attribute? Could not another attribute be invoked that is missing from `_df`?

ajkerrigan committed:
Make pandas sorting work with fewer tweaks

ajkerrigan commented:
I don't think that an attribute missing from `_df` is the problem, it's more the death spiral of trying to direct `__getattr__` to `self.df` when that hasn't been set yet. So another way to handle that would be to raise that `AttributeError` from `__getattr__` when `self.df` doesn't exist yet. That's closer to my original change, the difference is that we don't really care which attribute we're looking for (we're not checking `k`). We just know that `__getattr__` is a fallback lookup, and we're making sure `self.df` exists before we look there. Maybe that's less weird than the custom state logic?

anjakefala commented:
Looks great! Merging in. =) Thanks *so* much, @ajkerrigan!
